src/core/pagerankGraph.js

```diff
@@ -286,6 +299,88 @@ export class PagerankGraph {
     return weight;
   }
 
+  /**
+   * Provides the Neighbors to a target node, along with how those
+   * neighbors contributed to the node's score.
+   *
+   * See the docs on `Graph.neighbors` for the semantics of what a `Neighbor`
+   * is. This call augments the Neighbors from graph, so that for each neighbor
+   * we also have the neighbor node's score, the EdgeWeight for the edge, and a
+   * scoreContribution, which shows how much score was contributed to the
+   * target node from that Neighbor.
+   *
+   * When the PagerankGraph is well-converged, it will be the case that a
+   * node's score is equal to the score contribution from each neighbor plus
+   * the synthetic loop's score contribution.
+   *
+   * When the PagerankGraph is not well-converged, the score contributions are
+   * meaningless.
+   */
+  neighbors(
+    target: NodeAddressT,
+    options: NeighborsOptions
+  ): Iterator<ScoredNeighbor> {
+    this._verifyGraphNotModified();
+    if (!this.graph().hasNode(target)) {
+      throw new Error(
+        `Tried to find neighbors of non-existent node ${NodeAddress.toString(
+          target
+        )}`
+      );
+    }
+    return this._neighborsIterator(target, options);
+  }
+
+  *_neighborsIterator(
+    target: NodeAddressT,
+    options: NeighborsOptions
+  ): Iterator<ScoredNeighbor> {
+    const graphNeighbors = this.graph().neighbors(target, options);
+    for (const {node, edge} of graphNeighbors) {
+      const scoredNode = NullUtil.get(this.node(node));
+      const weightedEdge = NullUtil.get(this.edge(edge.address));
+      // We compute how much of target's score is attributable to the neighbor.
+      // First, we find out how much edge weight there was from node to target,
+      // based on whether it was an IN-edge or OUT-edge or loop.
+      let relevantEdgeWeight = 0;
+      if (edge.src === target) {
+        relevantEdgeWeight += weightedEdge.weight.froWeight;
+      }
+      if (edge.dst === target) {
+        relevantEdgeWeight += weightedEdge.weight.toWeight;
+      }
+      // We normalize this edge weight by the total outWeight for `node`.
+      const normalizedEdgeWeight =
+        relevantEdgeWeight / this.totalOutWeight(node);
```

Comment:
If you ask for outgoing edges in `options`, do you still want to divide those weights by `totalOutWeight` of `node`?Wouldn't you be dividing `node`'s incoming edge weight by its `totalOutWeight`? And the total incoming edge weight does not equal the total outgoing edge weight? (but total incoming score does equal total outgoing score, doesn't it?? Which sort of makes this difficult to model on a napkin)

So maybe if you are asking for outgoing edges, divide by the `totalOutWeight` of `target`? So if you ask for incoming edges, the `scoredContribution` is `node` 's contribution to `target`, and if you ask for outgoing edges, you get `target`'s contribution to `node` ?

Reply:
Remember that for the purposes of pagerank/the weighted graph, every edge is actually two edges: one pointing from src to dst, and one pointing from dst to src. So even if we are filtering for outgoing edges (i.e. where target is the src), we still have an edge coming back from `node` to `target`. Then to get the normalized weight of the edge from `node` to `target`, we need to normalize by `node`'s total weight (because the markov chain rule is that the weight of all the connections out of a node need to sum to one).

Basically, the concept of "outgoing edge" at the neighbor API level is orthogonal to the concept of "outgoing connection", since every edge (whether inbound or outbound) corresponds to two connections.

Does that make sense?

Reply:
I got `froWeight` and `toWeight` mixed up. I was reading the following snippet as accumulating weights from `target` to `node` instead of `node` to `target`: 
```javascript 
 if (edge.src === target) {
        relevantEdgeWeight += weightedEdge.weight.froWeight;
  }
```